File: kattis/easy/dicecup.java
package easy;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class dicecup {
	public static void main(String[] args) {
		try (Scanner scanner = new Scanner(System.in)) {
			int n = scanner.nextInt();
			int m = scanner.nextInt();

			assert 4 <= n && n <= 20;
			assert 4 <= m && m <= 20;

			HashMap<Integer, Integer> diceCounts = new HashMap<>();

			// first dice roll
			for (int i=1; i <= n; i++) {

				// second dice rolle
				for (int j=1; j <= m; j++) {

					// calculates sum of both throws and saves it
					int sum = i + j;
					diceCounts.put(sum, diceCounts.getOrDefault(sum, 0)+1);
				}
			}


			int highestCount = Integer.MIN_VALUE;
			ArrayList<Integer> highestKeys = new ArrayList<>();

			for (Map.Entry<Integer, Integer> entry : diceCounts.entrySet()) {
				Integer key = entry.getKey();
				Integer val = entry.getValue();

				if (val > highestCount) {
					highestCount = val;
					highestKeys.clear(); // Clear previous highest keys
					highestKeys.add(key);
				} else if (val == highestCount) {
					highestKeys.add(key); // Add another key with the same highest value
				}
			}

			highestKeys.forEach(System.out::println);
		}
	}
}

File: kattis/easy/ccpp/dicecup.cpp
//
// Created by BBJ on 08.01.2024.
//
#include <iostream>
#include <map>
#include <list>

std::map<int,int> countDices(int &n, int &m);
std::list<int> getHighestKeys(std::map<int,int> &diceCount);


int main() {
    int n, m;
    std::cin >> n >> m;

    // counts dice throws
    std::map<int,int> diceCount = countDices(n, m);

    // gets highest throws
    std::list<int> highest = getHighestKeys(diceCount);

    // prints highest throws
    for (const auto &val : highest)
        std::cout << val << "\n";

    return 0;
}

std::map<int,int> countDices(int &n, int &m) {
    std::map<int, int> diceCount;

    // first dice roll
    for (int i=1; i <= n; ++i) {

        // second dice roll
        for (int j=1; j <= m; ++j) {

            // calculate sum of both throws and saves it
            int sum = i+j;
            diceCount[sum]++;
        }
    }
    return diceCount;
}

std::list<int> getHighestKeys(std::map<int,int> &diceCount) {
    int highestCount = 0;
    std::list<int> highestKeys;

    // go through map with throws
    for (const auto &pair : diceCount) {
        int key = pair.first;
        int val = pair.second;

        if  (val > highestCount) {
            highestCount = val;
            highestKeys.clear();
            highestKeys.push_back(key);
        }
        else if (val == highestCount) {
            highestKeys.push_back(key);
        }
    }
    return highestKeys;
}